File: cpp/StompClient.cpp
#include "StompClient.hpp"
#include <iostream>
#include <sstream>
#include <chrono>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <cstring>

// STOMP Client Implementation
StompClient::StompClient(const std::string& host, int port)
    : StompClient(host, port, "", "") {
}

StompClient::StompClient(const std::string& host, int port, const std::string& login, const std::string& passcode)
    : host_(host), port_(port), login_(login), passcode_(passcode), socket_(-1) {
}

StompClient::~StompClient() {
    shutdown();
}

std::future<void> StompClient::connect() {
    if (connected_ || connecting_) {
        std::promise<void> promise;
        promise.set_value();
        return promise.get_future();
    }
    
    connecting_ = true;
    
    return std::async(std::launch::async, [this]() {
        try {
            // Create socket
            socket_ = socket(AF_INET, SOCK_STREAM, 0);
            if (socket_ < 0) {
                throw std::runtime_error("Failed to create socket");
            }
            
            // Setup server address
            struct sockaddr_in serverAddr;
            std::memset(&serverAddr, 0, sizeof(serverAddr));
            serverAddr.sin_family = AF_INET;
            serverAddr.sin_port = htons(port_);
            
            if (inet_pton(AF_INET, host_.c_str(), &serverAddr.sin_addr) <= 0) {
                throw std::runtime_error("Invalid address/Address not supported");
            }
            
            // Connect to server
            if (::connect(socket_, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
                throw std::runtime_error("Connection failed");
            }
            
            // Send CONNECT frame
            auto connectFrame = StompFrame::connect(host_, login_, passcode_);
            sendFrame(*connectFrame);
            
            // Start frame reader
            shouldStop_ = false;
            frameProcessorThread_ = std::thread(&StompClient::readFrames, this);
            
            std::cout << "Connected to STOMP server at " << host_ << ":" << port_ << std::endl;
            
            // Wait for CONNECTED response
            auto startTime = std::chrono::steady_clock::now();
            while (!connected_ && connecting_) {
                auto now = std::chrono::steady_clock::now();
                auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(now - startTime);
                if (elapsed.count() > 5000) { // 5 second timeout
                    break;
                }
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
            }
            
            if (!connected_) {
                connecting_ = false;
                cleanup();
                throw std::runtime_error("Connection timeout - no CONNECTED frame received");
            }
            
        } catch (const std::exception& e) {
            connecting_ = false;
            notifyConnectionListener(false, "Connection failed: " + std::string(e.what()));
            throw;
        }
    });
}

void StompClient::disconnect() {
    if (!connected_) return;
    
    try {
        // Send DISCONNECT frame
        auto disconnectFrame = StompFrame::disconnect();
        sendFrame(*disconnectFrame);
        std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Give server time to process
    } catch (const std::exception& e) {
        // Log error but continue cleanup
        std::cerr << "Error during disconnect: " << e.what() << std::endl;
    }
    
    cleanup();
}

void StompClient::cleanup() {
    connected_ = false;
    connecting_ = false;
    shouldStop_ = true;
    
    if (socket_ >= 0) {
        ::close(socket_);
        socket_ = -1;
    }
    
    if (frameProcessorThread_.joinable()) {
        frameProcessorThread_.join();
    }
    
    {
        std::lock_guard<std::mutex> lock(subscriptionsMutex_);
        subscriptions_.clear();
    }
    
    notifyConnectionListener(false, "Disconnected");
}

std::string StompClient::subscribe(const std::string& destination, StompMessageListener listener) {
    if (!connected_) {
        throw std::runtime_error("Not connected to STOMP server");
    }
    
    std::string subscriptionId = "sub-" + std::to_string(subscriptionCounter_.fetch_add(1) + 1);
    auto subscription = std::make_unique<StompSubscription>(subscriptionId, destination, std::move(listener));
    
    {
        std::lock_guard<std::mutex> lock(subscriptionsMutex_);
        subscriptions_[subscriptionId] = std::move(subscription);
    }
    
    auto subscribeFrame = StompFrame::subscribe(destination, subscriptionId);
    sendFrame(*subscribeFrame);
    
    std::cout << "Subscribed to " << destination << " with subscription ID: " << subscriptionId << std::endl;
    
    return subscriptionId;
}

void StompClient::unsubscribe(const std::string& subscriptionId) {
    if (!connected_) return;
    
    std::unique_ptr<StompSubscription> subscription;
    {
        std::lock_guard<std::mutex> lock(subscriptionsMutex_);
        auto it = subscriptions_.find(subscriptionId);
        if (it != subscriptions_.end()) {
            subscription = std::move(it->second);
            subscriptions_.erase(it);
        }
    }
    
    if (subscription) {
        auto unsubscribeFrame = StompFrame::unsubscribe(subscriptionId);
        sendFrame(*unsubscribeFrame);
        
        std::cout << "Unsubscribed from " << subscription->destination << std::endl;
    }
}

void StompClient::send(const std::string& destination, const std::string& message) {
    if (!connected_) {
        throw std::runtime_error("Not connected to STOMP server");
    }
    
    auto sendFramePtr = StompFrame::send(destination, message);
    sendFrame(*sendFramePtr);
    
    std::cout << "Sent message to " << destination << ": " << message << std::endl;
}

void StompClient::sendFrame(const StompFrame& frame) {
    if (socket_ >= 0) {
        std::string serialized = frame.serialize();
        ssize_t sent = ::send(socket_, serialized.c_str(), serialized.length(), 0);
        if (sent < 0) {
            throw std::runtime_error("Failed to send frame");
        }
    }
}

void StompClient::readFrames() {
    try {
        while (!shouldStop_ && connected_) {
            std::string frameData = readFrame();
            if (!frameData.empty()) {
                auto frame = StompFrame::parse(frameData);
                if (frame) {
                    handleIncomingFrame(std::move(frame));
                }
            } else {
                break; // End of stream
            }
        }
    } catch (const std::exception& e) {
        if (connected_) {
            std::cerr << "Frame reading error: " << e.what() << std::endl;
            notifyConnectionListener(false, "Connection lost: " + std::string(e.what()));
        }
    }
    cleanup();
}

std::string StompClient::readFrame() {
    std::string frameData;
    char ch;
    
    while (!shouldStop_ && socket_ >= 0) {
        ssize_t received = recv(socket_, &ch, 1, 0);
        if (received <= 0) {
            break; // Connection closed or error
        }
        
        if (ch == '\0') { // Null terminator found
            return frameData;
        }
        frameData += ch;
    }
    
    return frameData.empty() ? "" : frameData;
}

void StompClient::handleIncomingFrame(std::unique_ptr<StompFrame> frame) {
    const std::string& command = frame->command();
    
    if (command == "CONNECTED") {
        handleConnected(*frame);
    } else if (command == "MESSAGE") {
        handleMessage(*frame);
    } else if (command == "RECEIPT") {
        handleReceipt(*frame);
    } else if (command == "ERROR") {
        handleError(*frame);
    } else {
        std::cout << "Received unknown frame: Command '" << command << "' not supported" << std::endl;
    }
}

void StompClient::handleConnected(const StompFrame& frame) {
    connected_ = true;
    connecting_ = false;
    
    auto it = frame.headers().find("session");
    if (it != frame.headers().end()) {
        sessionId_ = it->second;
    }
    
    std::cout << "Successfully connected to STOMP server. Session: " << sessionId_ << std::endl;
    notifyConnectionListener(true, "Connected");
}

void StompClient::handleMessage(const StompFrame& frame) {
    auto destIt = frame.headers().find("destination");
    auto subIt = frame.headers().find("subscription");
    
    if (destIt != frame.headers().end() && subIt != frame.headers().end()) {
        std::lock_guard<std::mutex> lock(subscriptionsMutex_);
        auto it = subscriptions_.find(subIt->second);
        if (it != subscriptions_.end() && it->second->listener) {
            try {
                it->second->listener(destIt->second, frame.body(), frame.headers());
            } catch (const std::exception& e) {
                std::cerr << "Error in message listener: " << e.what() << std::endl;
            }
        }
    }
}

void StompClient::handleReceipt(const StompFrame& frame) {
    auto it = frame.headers().find("receipt-id");
    if (it != frame.headers().end()) {
        std::cout << "Received receipt: " << it->second << std::endl;
    }
}

void StompClient::handleError(const StompFrame& frame) {
    auto it = frame.headers().find("message");
    std::string message = (it != frame.headers().end()) ? it->second : "Unknown error";
    
    std::cerr << "STOMP Error: " << message << std::endl;
    std::cerr << "Details: " << frame.body() << std::endl;
    
    if (errorListener_) {
        errorListener_(message, frame.body());
    }
}

void StompClient::setConnectionListener(StompConnectionListener listener) {
    connectionListener_ = std::move(listener);
}

void StompClient::setErrorListener(StompErrorListener listener) {
    errorListener_ = std::move(listener);
}

void StompClient::notifyConnectionListener(bool connected, const std::string& reason) {
    if (connectionListener_) {
        try {
            connectionListener_(connected, reason);
        } catch (const std::exception& e) {
            std::cerr << "Error in connection listener: " << e.what() << std::endl;
        }
    }
}

std::set<std::string> StompClient::getActiveSubscriptions() const {
    std::lock_guard<std::mutex> lock(subscriptionsMutex_);
    std::set<std::string> result;
    for (const auto& [id, subscription] : subscriptions_) {
        result.insert(id);
    }
    return result;
}

void StompClient::shutdown() {
    disconnect();
    if (frameProcessorThread_.joinable()) {
        frameProcessorThread_.join();
    }
}

void StompClient::runTestClient(const std::string& host, int port) {
    StompClient client(host, port);
    
    // Set up listeners
    client.setConnectionListener([](bool connected, const std::string& reason) {
        std::cout << "Connection status: " << connected << " - " << reason << std::endl;
    });
    
    client.setErrorListener([](const std::string& message, const std::string& details) {
        std::cerr << "STOMP Error: " << message << "\nDetails: " << details << std::endl;
    });
    
    try {
        // Connect to server
        client.connect().get();
        
        // Subscribe to a test topic
        std::string subscription = client.subscribe("/topic/test", 
            [](const std::string& destination, const std::string& body, const std::map<std::string, std::string>& /* headers */) {
                std::cout << "Received message on " << destination << ": " << body << std::endl;
            });
        
        // Interactive console
        std::cout << "STOMP Client started. Commands:" << std::endl;
        std::cout << "  send <destination> <message> - Send a message" << std::endl;
        std::cout << "  subscribe <destination> - Subscribe to destination" << std::endl;
        std::cout << "  unsubscribe <subscription-id> - Unsubscribe" << std::endl;
        std::cout << "  quit - Exit" << std::endl;
        
        std::string input;
        while (client.isConnected() && std::getline(std::cin, input)) {
            if (input.empty()) continue;
            
            if (input == "quit") {
                break;
            }
            
            std::istringstream iss(input);
            std::string command;
            iss >> command;
            
            try {
                if (command == "send") {
                    std::string destination, message;
                    iss >> destination;
                    std::getline(iss, message);
                    if (!message.empty() && message[0] == ' ') {
                        message = message.substr(1); // Remove leading space
                    }
                    
                    if (!destination.empty() && !message.empty()) {
                        client.send(destination, message);
                    } else {
                        std::cout << "Usage: send <destination> <message>" << std::endl;
                    }
                } else if (command == "subscribe") {
                    std::string destination;
                    iss >> destination;
                    
                    if (!destination.empty()) {
                        std::string subId = client.subscribe(destination, 
                            [destination](const std::string& dest, const std::string& body, const std::map<std::string, std::string>& /* headers */) {
                                std::cout << "Message on " << dest << ": " << body << std::endl;
                            });
                        std::cout << "Subscribed with ID: " << subId << std::endl;
                    } else {
                        std::cout << "Usage: subscribe <destination>" << std::endl;
                    }
                } else if (command == "unsubscribe") {
                    std::string subscriptionId;
                    iss >> subscriptionId;
                    
                    if (!subscriptionId.empty()) {
                        client.unsubscribe(subscriptionId);
                    } else {
                        std::cout << "Usage: unsubscribe <subscription-id>" << std::endl;
                    }
                } else if (command == "status") {
                    std::cout << "Connected: " << client.isConnected() << std::endl;
                } else {
                    std::cout << "Unknown command: " << command << std::endl;
                }
            } catch (const std::exception& e) {
                std::cerr << "Error: " << e.what() << std::endl;
            }
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Client error: " << e.what() << std::endl;
    }
    
    client.shutdown();
}

File: ts/src/StompClient.ts
import * as net from 'net';
import { StompFrame, StompFrameUtils } from './StompFrame.js';

// Listener interfaces
export type StompMessageListener = (destination: string, body: string, headers: Record<string, string>) => void;
export type StompConnectionListener = (connected: boolean, reason: string) => void;
export type StompErrorListener = (message: string, details: string) => void;

// Subscription holder
export interface StompSubscription {
  subscriptionId: string;
  destination: string;
  listener: StompMessageListener;
}

// STOMP Client Implementation
export class StompClient {
  private readonly host: string;
  private readonly port: number;
  private readonly login?: string;
  private readonly passcode?: string;
  
  private socket?: net.Socket;
  private connected = false;
  private connecting = false;
  private sessionId?: string;
  
  // Listeners
  private connectionListener?: StompConnectionListener;
  private errorListener?: StompErrorListener;
  
  // Subscriptions management
  private readonly subscriptions = new Map<string, StompSubscription>();
  private subscriptionCounter = 0;
  
  constructor(host: string, port: number, login?: string, passcode?: string) {
    this.host = host;
    this.port = port;
    this.login = login;
    this.passcode = passcode;
  }
  
  // Connection management
  async connect(): Promise<void> {
    if (this.connected || this.connecting) {
      return;
    }
    
    this.connecting = true;
    
    return new Promise((resolve, reject) => {
      try {
        this.socket = new net.Socket();
        
        this.socket.connect(this.port, this.host, () => {
          console.log(`Connected to STOMP server at ${this.host}:${this.port}`);
          
          // Send CONNECT frame
          const connectFrame = StompFrameUtils.connect(this.host, this.login, this.passcode);
          this.sendFrame(connectFrame);
          
          // Start frame reader
          this.readFrames();
          
          // Wait for CONNECTED response
          const startTime = Date.now();
          const checkConnection = () => {
            if (this.connected) {
              resolve();
            } else if (this.connecting && (Date.now() - startTime) < 5000) {
              setTimeout(checkConnection, 10);
            } else {
              this.connecting = false;
              this.cleanup();
              reject(new Error('Connection timeout - no CONNECTED frame received'));
            }
          };
          checkConnection();
        });
        
        this.socket.on('error', (err) => {
          this.connecting = false;
          this.notifyConnectionListener(false, `Connection failed: ${err.message}`);
          reject(new Error(`Failed to connect to STOMP server: ${err.message}`));
        });
        
      } catch (error) {
        this.connecting = false;
        this.notifyConnectionListener(false, `Connection failed: ${error}`);
        reject(new Error(`Failed to connect to STOMP server: ${error}`));
      }
    });
  }
  
  disconnect(): void {
    if (!this.connected) return;
    
    try {
      // Send DISCONNECT frame
      this.sendFrame(StompFrameUtils.disconnect());
      setTimeout(() => this.cleanup(), 100); // Give server time to process
    } catch (error) {
      // Log error in production
      this.cleanup();
    }
  }
  
  private cleanup(): void {
    this.connected = false;
    this.connecting = false;
    
    if (this.socket && !this.socket.destroyed) {
      this.socket.destroy();
    }
    
    this.subscriptions.clear();
    this.notifyConnectionListener(false, 'Disconnected');
  }
  
  // Subscription management
  subscribe(destination: string, listener: StompMessageListener): string {
    if (!this.connected) {
      throw new Error('Not connected to STOMP server');
    }
    
    const subscriptionId = `sub-${++this.subscriptionCounter}`;
    const subscription: StompSubscription = {
      subscriptionId,
      destination,
      listener
    };
    
    this.subscriptions.set(subscriptionId, subscription);
    
    const subscribeFrame = StompFrameUtils.subscribe(destination, subscriptionId);
    this.sendFrame(subscribeFrame);
    
    console.log(`Subscribed to ${destination} with subscription ID: ${subscriptionId}`);
    
    return subscriptionId;
  }
  
  unsubscribe(subscriptionId: string): void {
    if (!this.connected) return;
    
    const subscription = this.subscriptions.get(subscriptionId);
    if (subscription) {
      this.subscriptions.delete(subscriptionId);
      const unsubscribeFrame = StompFrameUtils.unsubscribe(subscriptionId);
      this.sendFrame(unsubscribeFrame);
      
      console.log(`Unsubscribed from ${subscription.destination}`);
    }
  }
  
  // Message sending
  send(destination: string, message: string): void {
    if (!this.connected) {
      throw new Error('Not connected to STOMP server');
    }
    
    const sendFrame = StompFrameUtils.send(destination, message);
    this.sendFrame(sendFrame);
    
    console.log(`Sent message to ${destination}: ${message}`);
  }
  
  // Frame processing
  private sendFrame(frame: StompFrame): void {
    if (this.socket && !this.socket.destroyed) {
      const serialized = StompFrameUtils.serialize(frame);
      this.socket.write(serialized, 'utf8');
    }
  }
  
  private readFrames(): void {
    if (!this.socket) return;
    
    let buffer = '';
    
    this.socket.on('data', (data) => {
      buffer += data.toString('utf8');
      
      // Process complete frames (ending with null terminator)
      let nullIndex;
      while ((nullIndex = buffer.indexOf('\0')) !== -1) {
        const frameData = buffer.substring(0, nullIndex);
        buffer = buffer.substring(nullIndex + 1);
        
        if (frameData.length > 0) {
          const frame = StompFrameUtils.parse(frameData);
          if (frame) {
            this.handleIncomingFrame(frame);
          }
        }
      }
    });
    
    this.socket.on('close', () => {
      if (this.connected) {
        console.error('Connection lost');
        this.notifyConnectionListener(false, 'Connection lost');
      }
      this.cleanup();
    });
    
    this.socket.on('error', (err) => {
      if (this.connected) {
        console.error(`Frame reading error: ${err.message}`);
        this.notifyConnectionListener(false, `Connection lost: ${err.message}`);
      }
      this.cleanup();
    });
  }
  
  private handleIncomingFrame(frame: StompFrame): void {
    switch (frame.command) {
      case 'CONNECTED':
        this.handleConnected(frame);
        break;
      case 'MESSAGE':
        this.handleMessage(frame);
        break;
      case 'RECEIPT':
        this.handleReceipt(frame);
        break;
      case 'ERROR':
        this.handleError(frame);
        break;
      default:
        console.log(`Received unknown frame: Command '${frame.command}' not supported`);
    }
  }
  
  private handleConnected(frame: StompFrame): void {
    this.connected = true;
    this.connecting = false;
    this.sessionId = frame.headers['session'];
    
    console.log(`Successfully connected to STOMP server. Session: ${this.sessionId}`);
    this.notifyConnectionListener(true, 'Connected');
  }
  
  private handleMessage(frame: StompFrame): void {
    const destination = frame.headers['destination'];
    const subscriptionId = frame.headers['subscription'];
    
    if (destination && subscriptionId) {
      const subscription = this.subscriptions.get(subscriptionId);
      if (subscription && subscription.listener) {
        try {
          subscription.listener(destination, frame.body, frame.headers);
        } catch (error) {
          console.error(`Error in message listener: ${error}`);
        }
      }
    }
  }
  
  private handleReceipt(frame: StompFrame): void {
    const receiptId = frame.headers['receipt-id'];
    console.log(`Received receipt: ${receiptId}`);
  }
  
  private handleError(frame: StompFrame): void {
    const message = frame.headers['message'];
    console.error(`STOMP Error: ${message}`);
    console.error(`Details: ${frame.body}`);
    
    if (this.errorListener) {
      this.errorListener(message || 'Unknown error', frame.body);
    }
  }
  
  // Listeners
  setConnectionListener(listener: StompConnectionListener): void {
    this.connectionListener = listener;
  }
  
  setErrorListener(listener: StompErrorListener): void {
    this.errorListener = listener;
  }
  
  private notifyConnectionListener(connected: boolean, reason: string): void {
    if (this.connectionListener) {
      try {
        this.connectionListener(connected, reason);
      } catch (error) {
        console.error(`Error in connection listener: ${error}`);
      }
    }
  }
  
  // Status methods
  isConnected(): boolean {
    return this.connected;
  }
  
  getSessionId(): string | undefined {
    return this.sessionId;
  }
  
  getActiveSubscriptions(): Set<string> {
    return new Set(this.subscriptions.keys());
  }
  
  // Shutdown
  shutdown(): void {
    this.disconnect();
  }
}
